pkg/remotecmd/kubeexec.go

```diff
@@ -0,0 +1,294 @@
+		cmd = append(cmd, fmt.Sprintf("%s && (%s) %s", pidWriterCmd, cmdLine, outputRedirectCmd))
+	}
+
+	go func() {
+		if outputHandler != nil {
+			outputHandler(Starting, nil, nil, nil)
+		}
+
+		stdout, stderr, err := ExecuteCommandAndGetOutput(kclient, podName, containerName, false, cmd...)
+		if err != nil {
+			klog.V(2).Infof("error while running background command: %v", err)
+		}
+
+		if outputHandler != nil {
+			outputHandler(Stopped, stdout, stderr, err)
+		}
+	}()
+
+	return nil
+}
+
+// StopProcessForCommand stops the process representing the specified Devfile command.
+// Because of the way this process is launched and its PID stored (see StartProcessForCommand),
+// we need to determine the process children (there should be only one child). Then killing those children
+// will exit the parent 'sh' process.
+func (k *kubeExecProcessHandler) StopProcessForCommand(
+	devfileCmd devfilev1.Command,
+	kclient kclient.ClientInterface,
+	podName string,
+	containerName string,
+) error {
+	klog.V(4).Infof("StopProcessForCommand for %q", devfileCmd.Id)
+	defer func() {
+		pidFile := getPidFileForCommand(devfileCmd)
+		err := ExecuteCommand([]string{common.ShellExecutable, "-c", fmt.Sprintf("rm -f %s", pidFile)},
+			kclient, podName, containerName, false, nil, nil)
+		if err != nil {
+			klog.V(2).Infof("Could not remove file %q: %v", pidFile, err)
+		}
+	}()
+
+	ppid, err := getRemoteProcessPID(kclient, devfileCmd, podName, containerName)
+	if err != nil {
+		return err
+	}
+	if ppid == 0 {
+		return nil
+	}
+
+	children, err := getProcessChildren(ppid, kclient, podName, containerName)
+	if err != nil {
+		return err
+	}
+
+	kill := func(p int) error {
+		err = ExecuteCommand([]string{common.ShellExecutable, "-c", fmt.Sprintf("kill %d || true", p)},
+			kclient, podName, containerName, false, nil, nil)
+
+		//Because the process we just stopped might take longer to exit (it might have caught the signal and is performing additional cleanup),
+		//retry detecting its actual state till it is stopped or timeout expires
+		var processInfo interface{}
+		processInfo, err = task.NewRetryable(fmt.Sprintf("status for remote process %d", p), func() (bool, interface{}, error) {
+			pInfo, e := k.getProcessInfoFromPid(p, kclient, podName, containerName)
+			return e == nil || pInfo.Status == Stopped, pInfo, e
+		}, true).RetryWithSchedule(2*time.Second, 4*time.Second, 8*time.Second)
+		if err != nil {
+			return err
+		}
+
+		pInfo, ok := processInfo.(RemoteProcessInfo)
+		if !ok {
+			klog.V(2).Infof("invalid type for remote process (%d) info, expected RemoteProcessInfo", p)
+			return fmt.Errorf("internal error while checking remote process status: %d", p)
+		}
+		if pInfo.Status != Stopped {
+			return fmt.Errorf("invalid status for remote process %d: %+v", p, processInfo)
+		}
+		return nil
+	}
+
+	if len(children) == 0 {
+		return kill(ppid)
+	}
```

Comment:
I don't see why you need to do a separate check for `len(children)`, I think using the for loop instead would be fine and not make much difference. Also, if there should only be one child, shouldn't we check that here and raise an error if it's not true?

Reply:
> I don't see why you need to do a separate check for `len(children)`, I think using the for loop instead would be fine and not make much difference. 
>
I wanted to fall back to killing the parent process if there were no identified children. But you're right, this could have been done with a single loop iterating over both the parent and all the children PIDs found. I'll change that.

> Also, if there should only be one child, shouldn't we check that here and raise an error if it's not true?
>
For now, there is one child, but I wanted to be safe if we later change the final command-line and end up with several children. It was cheap to add, so I wanted to be able to handle that as well.